protocol/daemons/server/types/health_checker.go

```diff
@@ -0,0 +1,169 @@
+package types
+
+import (
+	"github.com/cometbft/cometbft/libs/log"
+	"github.com/dydxprotocol/v4-chain/protocol/daemons/types"
+	libtime "github.com/dydxprotocol/v4-chain/protocol/lib/time"
+	"sync"
+	"time"
+)
+
+// timestampWithError couples a timestamp and error to make it easier to update them in tandem. The
+// timestampWithError will track the timestamp of the first error in a streak of errors, but keeps a record of the
+// most recent error. This is useful for determining how long a service has been unhealthy, as well as the current
+// state of the service.
+type timestampWithError struct {
+	timestamp time.Time
+	err       error
+}
+
+func (u *timestampWithError) Update(timestamp time.Time, err error) {
```

Comment:
Since `Update` suggests that both values will be updated, should update this to something that hints that it only updates err if the timestamp is set?

```suggestion
func (u *timestampWithError) UpdateLastErr(timestamp time.Time, err error) {
```